sdk/src/azure/iot/az_iot_adu_ota.c

```diff
@@ -0,0 +1,899 @@
+        RETURN_IF_JSON_TOKEN_TYPE_NOT(jr, AZ_JSON_TOKEN_STRING);
+
+        required_size = jr->token.size + NULL_TERM_CHAR_SIZE;
+
+        _az_RETURN_IF_NOT_ENOUGH_SIZE(buffer, required_size);
+
+        update_request->file_urls[update_request->file_urls_count].url
+            = split_az_span(buffer, required_size, &buffer);
+
+        _az_RETURN_IF_FAILED(az_json_token_get_string(
+            &jr->token,
+            (char*)az_span_ptr(update_request->file_urls[update_request->file_urls_count].url),
+            az_span_size(update_request->file_urls[update_request->file_urls_count].url),
+            &out_length));
+
+        // TODO: find a way to get rid of az_json_token_get_string (which adds a \0 at the
+        // end!!!!!!)
+        //       Preferably have a function that does not copy anything.
+        update_request->file_urls[update_request->file_urls_count].url = az_span_slice(
+            update_request->file_urls[update_request->file_urls_count].url, 0, out_length);
+
+        update_request->file_urls_count++;
+
+        _az_RETURN_IF_FAILED(az_json_reader_next_token(jr));
+      }
+    }
+
+    _az_RETURN_IF_FAILED(az_json_reader_next_token(jr));
+  }
+
+  if (buffer_remainder != NULL)
+  {
+    *buffer_remainder = buffer;
+  }
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_iot_adu_ota_get_service_properties_response(
+    az_iot_hub_client const* iot_hub_client,
+    int32_t version,
+    int32_t status,
+    az_span payload,
+    az_span* out_payload)
+{
+  _az_PRECONDITION_NOT_NULL(iot_hub_client);
+  _az_PRECONDITION_VALID_SPAN(payload, 1, false);
+  _az_PRECONDITION_NOT_NULL(out_payload);
+
+  az_json_writer jw;
+
+  // Component and response status
+  _az_RETURN_IF_FAILED(az_json_writer_init(&jw, payload, NULL));
+  _az_RETURN_IF_FAILED(az_json_writer_append_begin_object(&jw));
+  _az_RETURN_IF_FAILED(az_iot_hub_client_properties_writer_begin_component(
+      iot_hub_client, &jw, AZ_SPAN_FROM_STR(AZ_IOT_ADU_OTA_AGENT_COMPONENT_NAME)));
+  _az_RETURN_IF_FAILED(az_iot_hub_client_properties_writer_begin_response_status(
+      iot_hub_client,
+      &jw,
+      AZ_SPAN_FROM_STR(AZ_IOT_ADU_OTA_AGENT_PROPERTY_NAME_SERVICE),
+      status,
+      version,
+      AZ_SPAN_EMPTY));
+
+  // It is not necessary to send the properties back in the acknowledgement.
+  // We opt not to send them to reduce the size of the payload.
+  _az_RETURN_IF_FAILED(az_json_writer_append_begin_object(&jw));
+  _az_RETURN_IF_FAILED(az_json_writer_append_end_object(&jw));
+
+  _az_RETURN_IF_FAILED(
+      az_iot_hub_client_properties_writer_end_response_status(iot_hub_client, &jw));
+  _az_RETURN_IF_FAILED(az_iot_hub_client_properties_writer_end_component(iot_hub_client, &jw));
+  _az_RETURN_IF_FAILED(az_json_writer_append_end_object(&jw));
+
+  *out_payload = az_json_writer_get_bytes_used_in_destination(&jw);
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_iot_adu_ota_parse_update_manifest(
```

Comment:
This function is ~300 lines long.  Please refactor - even if it's only into helpers that are called once - to keep high-level flow easier with fewer levels of indentation.